### PR TITLE
Implement new REPL for Natalie

### PR DESCRIPTION
**Pull request by TheGrizzlyDev**
This new REPL should:
- Allow multiline editing (done)
- Allow highlighting (wip)
- Have a navigable history (done)
- Look good
- Profit (?)

In order to enable it you will need to explicitly pass the option --experimental-repl-v2 and even then it will only be considered if natalie isn't self-hosted.

image

### CODE REVIEW

File: bin/natalie [hunk 1]
```diff
@@ -100,7 +104,7 @@ class Runner
   def run
     load_code
     if @repl
-      repl = Natalie::Repl.new
+      repl = if options[:experimental_repl_v2] then Natalie::ExperimentalReplV2.new else Natalie::Repl.new end
```
Comment (seven1m) [inline bin/natalie:107]: Let's use a ternary here...

```suggestion
      repl = options[:experimental_repl_v2] ? Natalie::ExperimentalReplV2.new : Natalie::Repl.new
```

File: lib/natalie/repl/repl.rb [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+    @   N N       N   @    A  A        T        A  A    L       I    E
+    @   N N N     N   @   A    A       T       A    A   L       I    EEE
+    @   N     N   N   @   AAAAAA       T       AAAAAA   L       I    E
+    @   N       N N   @   A    A       T       A    A   LLLLL   I    EEEEE
+    @                 @ 
+    @@@@@@@@@@@@@@@@@@@
+    \ \ \ \ \ \ \ :::
+    @@@@@@@@@@@@@@@@@@@#{RESET_STYLE_ASCCI_CODE}
+    \n
+    "
```
Comment (seven1m) [inline lib/natalie/repl/repl.rb:51]: What do you think about using those braille characters to generate something a bit more like the actual logo...

```
  ⡎⠉⣉⣉⣉⣉⣉⣉⣉⣉⠉⢱
  ⡇⢸ ⡢⡢⡀ ⠠⡢ ⡇⢸
  ⡇⢸ ⡪⡊⠪⡢⡨⡪ ⡇⢸
  ⡇⢸ ⠊⠂ ⠈⠊⠊ ⡇⢸
  ⣇⣀⣉⣉⣉⣉⣉⣉⣉⣉⣀⣸
⡔⣒⠒⠒⠒⠒⠚⠒⠒⠓⠒⠒⠒⠒⠒⢢
⢇⣉⣀⣀⣀⣀⣀⣀⣀⣀⣁⣉⣉⣉⣁⡸
```

Note: GitHub has some alignment issues here, but this seems to work in the terminal for me. Would love to see if it works in your terminal too.

Comment (TheGrizzlyDev) [reply lib/natalie/repl/repl.rb:51]: it's wonderful and it's all I ever wanted :^) Out of curiosity, how long did it take?

Comment (seven1m) [reply lib/natalie/repl/repl.rb:51]: Didn't take too long. I used this tool: https://braille-txt.github.io/ :wink:

**Comment by seven1m** [comment]
This is pretty sweet. There are some issues with it crashing and getting confused about multi-line expressions easily. I don't know if we want to try to fix those issues before we merge, or just go ahead and merge and iterate later.

File: lib/natalie/repl/repl.rb (hunk 2 above)
Comment (TheGrizzlyDev) [reply lib/natalie/repl/repl.rb:51]: This is cheating *coff coff*

**Comment by TheGrizzlyDev** [comment]
> This is pretty sweet. There are some issues with it crashing and getting confused about multi-line expressions easily. I don't know if we want to try to fix those issues before we merge, or just go ahead and merge and iterate later.

I'd merge it and give it a chance for others to have fun with, it's under an experimental flag anyway :)

**Comment by TheGrizzlyDev** [comment]
@seven1m done, the ascii art looks so good!

**Pull request merged**